### PR TITLE
Add settings modal

### PR DESCRIPTION
**Pull request by danwang**
r? @rbtying 
Move settings into a modal.

I also changed the data flow to not create a new React root every time settings change by pulling state out into SettingsProvider

There's a button on the top now:
Screen Shot 2020-04-05 at 7 17 30 PM

Clicking it opens this modal:
Screen Shot 2020-04-05 at 7 17 35 PM

### CODE REVIEW

File: frontend/src/DivWithProps.tsx [hunk 1]
```diff
@@ -0,0 +1,10 @@
+import * as React from 'react';
+
+type Props = {
+  children: React.ReactNode;
+};
+const DivWithProps = (extraProps: {}) => (props: Props) => (
```
Comment (rbtying) [inline frontend/src/DivWithProps.tsx:6]: huh. what is the purpose of this?

File: frontend/src/icons/Gear.tsx [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+const Gear = ({width = '100%', fill = '#000'}: Props) => (
+  <svg
+    focusable="false"
+    role="img"
+    xmlns="http://www.w3.org/2000/svg"
+    viewBox="0 0 512 512"
+    width={width}
+  >
+    <path
+      fill={fill}
```
Comment (rbtying) [inline frontend/src/icons/Gear.tsx:16]: is this an inline gear svg...

this is amaze

File: frontend/src/DivWithProps.tsx (hunk 1 above)
Comment (danwang) [reply frontend/src/DivWithProps.tsx:6]: it's a component maker for the times when you just need a div with some CSS on it, but don't want to give it a name/file

File: frontend/src/icons/Gear.tsx (hunk 2 above)
Comment (danwang) [reply frontend/src/icons/Gear.tsx:16]: I couldn't figure out how to import non-JS so everything is just JS 🙈

**Pull request merged**